### PR TITLE
Add schema for engine configs, improve documentation

### LINKED ISSUE #74: Support schema for engine config and recipes
https://json-schema.org/

### PR DESCRIPTION
**Pull request by umairidris**
#74

### CODE REVIEW

File: internal/tfengine/schema.go [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+        recipe_path:
+          description: Path to a recipe YAML config. Mutually exclusive with 'component_path'.
+          type: string
+
+        output_ref:
+          description: |
+            OutputRef allows this template write its contents to the same dir as another template.
+
+            Example:
+            If output_ref is set to 'foo.bar' then this will look for sibling template 'foo' (defined in this list) and a child template inside 'foo' called 'bar'.
```
Comment (MartinPetkov) [inline internal/tfengine/schema.go:51]: I'm still unclear on how `output_ref` functions. How does the child template 'bar' get used in this case?

File: internal/tfengine/schema.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+          type: string
+
+        data:
```
Comment (MartinPetkov) [inline internal/tfengine/schema.go:54]: Does this need a type, i.e. type: object? Or is that the default?

File: internal/tfengine/schema.go [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+            Example:
+              - Flag --output_path = /tmp/engine
+              - Field 'output_ref' set to a template which set output_path to './foo'
+              - This field is set to './bar'
+              - The final output path for this template is: '/tmp/engine/foo/bar'.
+          type: string
+
+        flatten:
+          description: |
```
Comment (MartinPetkov) [inline internal/tfengine/schema.go:81]: Does flatten recursively flatten? Or only one level? i.e. if the AUDIT key's value was an object, would its keys be flattened too.

File: internal/tfengine/load.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+	if err := json.Unmarshal(cj, c); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+func validate(confJSON []byte) error {
+	schemaJSON, err := yaml.YAMLToJSON([]byte(schema))
+	if err != nil {
+		return fmt.Errorf("failed to load schema: %v", err)
```
Comment (xingao267) [inline internal/tfengine/load.go:58]: remove "failed to" as in v3 we are trying to use just something like fmt.Errorf("load schema")

File: internal/tfengine/load.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+	}
+
+	c := new(Config)
+	if err := json.Unmarshal(cj, c); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+func validate(confJSON []byte) error {
```
Comment (xingao267) [inline internal/tfengine/load.go:55]: Can some of these logic be moved to template package and can later be reused by policygen?

File: internal/tfengine/schema.go (hunk 1 above)
Comment (umairidris) [reply internal/tfengine/schema.go]: I added more explanation, let me know if this is clearer. There will also be schemas for each recipe that will list all the output_refs for each recipe.

File: internal/tfengine/schema.go (hunk 3 above)
Comment (umairidris) [reply internal/tfengine/schema.go:88]: Only one level, clarified. Good question.

File: internal/tfengine/load.go (hunk 5 above)
Comment (umairidris) [reply internal/tfengine/load.go]: Done

File: internal/tfengine/load.go (hunk 4 above)
Comment (umairidris) [reply internal/tfengine/load.go]: Done

File: internal/tfengine/schema.go [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+          type: string
+
+        recipe_path:
+          description: Path to a recipe YAML config. Mutually exclusive with 'component_path'.
+          type: string
+
+        output_ref:
+          description: |
+            Every template outputs to a specific directory. This field allows
+            a template write its contents to the same output dir as another
```
Comment (MartinPetkov) [inline internal/tfengine/schema.go:49]: Typo, "**to** write"

Comment (umairidris) [reply internal/tfengine/schema.go]: Done

File: internal/jsonschema/validate.go [hunk 7]
```diff
@@ -0,0 +1,28 @@
+// Package jsonschema defines helpers for JSON schema validation.
```
Comment (xingao267) [inline internal/jsonschema/validate.go:1]: license

Comment (umairidris) [reply internal/jsonschema/validate.go:15]: Done

**Pull request merged**